### PR TITLE
[121, 123, 125] Place a bid (Backend), Re-bid on the lot (Backend), Withdraw a bid (Backend), respectively

### PR DESCRIPTION
**Pull request by khasanrashidov**
### Where

[NL-121 Place a bid (Backend)](https://jjbcapital.atlassian.net/browse/NL-121)
[NL-123 Re-bid on the lot (Backend)](https://jjbcapital.atlassian.net/browse/NL-123)
[NL-125 Withdraw a bid (Backend)](https://jjbcapital.atlassian.net/browse/NL-125)

### What

Developing the endpoint for bidding, rebidding, and bid withdrawal. Additionally, implementing (developing) the get all bids for buyer endpoint.

### How

Implemented all necessary endpoints for bids Place a bid (Backend), Re-bid on the lot (Backend), Withdraw a bid (Backend), View all user bids. To implement all of them, lots of changes were made, including in Db. Wrote 10 unit tests. (including for some other handlers as well).

### Affected Area 

Basically every domain was affected:

Auctionify.API.Controllers
Auctionify.Application
Auctionify.Core
Auctionify.Infrastructure

### PR Checklist

_Build_

- [x] Build Succeeded

_Unit Tests_

wrote new 10 unit tests (all passed)

- [x] All Tests Passed, 0 Ignored, 0 Failed

### CODE REVIEW

**Comment by khasanrashidov** [comment]
this is basically branch for ticket [118], but as tickets were connected to bids, the tickets [121, 123, 125] were implemented first and pushed into this branch.

**Comment by khasanrashidov** [comment]
SignalR was setup on seller lot profile (for now, so.. because the buyer lot profile is not developed yet (it is another ticket for other member), SignalR will be quickly set up without any efforts, when frontend for buyer lot profile will be developed.

To check the real time change (without page refreshing), you need to:

1. Sign In as a Seller in **frontend** and get the lot you want:
e.g. https://localhost:44452/seller/get-lot/2 

2. And you need to sign in as a Buyer in **backend** in swagger in order to bid for particular lot, when you bid, the page of seller lot profile will not refresh, but the new bid will be shown.

image



https://github.com/U6D5T4/Auctionify/assets/89701590/fcdec23e-8b07-426e-bd8c-ccfe90d314af

**Comment by khasanrashidov** [comment]
current look on signalR service (with seller view lot example) in frontend part:

image

**Comment by khasanrashidov** [comment]
withdraw bid

**Review by otash8110 — APPROVED** [review]

File: src/Auctionify.Application/Features/Users/Commands/AddBidForLot/AddBidForLotCommand.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+				BuyerId = user!.Id,
+				NewPrice = request.Bid,
+				TimeStamp = DateTime.Now,
+				LotId = request.LotId,
+				BidRemoved = false
+			};
+
+			var result = await _bidRepository.AddAsync(bid);
+
+			await _hubContext.Clients.All.SendAsync("ReceiveBidNotification", cancellationToken: cancellationToken);
```
Comment (SevenDart) [inline src/Auctionify.Application/Features/Users/Commands/AddBidForLot/AddBidForLotCommand.cs:60]: In current implementation of AuctionHub you send message to all clients that are looking at all lots, but you need to send message only to clients that are looking for specific lot from the request. To achieve that, you need to group your users by id. Check documentation of the groups feature: https://learn.microsoft.com/en-us/aspnet/core/signalr/groups?view=aspnetcore-8.0

File: src/Auctionify.Application/Features/Users/Commands/AddBidForLot/AddBidForLotCommandValidator.cs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+						}
+
+						return true;
+					}
+				)
+				.WithMessage("Bid must be greater than the lot's starting price")
+				.OverridePropertyName("Bid")
+				.WithName("Bid");
+
+			RuleFor(x => x)
```
Comment (SevenDart) [inline src/Auctionify.Application/Features/Users/Commands/AddBidForLot/AddBidForLotCommandValidator.cs:107]: Why do you need this check if user's new bid can't be lower than anyone else's(including this user) bid? What is the reason to check specially user bids?

File: src/Auctionify.Application/Features/Users/Commands/RemoveBid/RemoveBidCommand.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+			var bid = await _bidRepository.GetAsync(
+				predicate: x => x.Id == request.BidId,
+				cancellationToken: cancellationToken
+			);
+
+			bid!.BidRemoved = true;
+
+			await _bidRepository.UpdateAsync(bid!);
+
+			await _hubContext.Clients.All.SendAsync(
```
Comment (SevenDart) [inline src/Auctionify.Application/Features/Users/Commands/RemoveBid/RemoveBidCommand.cs:45]: Same as in AddBidForLotCommand

File: src/Auctionify.Application/Features/Users/Commands/RemoveBid/RemoveBidCommandValidator.cs
Comment (SevenDart) [inline src/Auctionify.Application/Features/Users/Commands/RemoveBid/RemoveBidCommandValidator.cs:1]: I think you also need to check that withdrawing bid was performed by the user that is requiring for withdraw

File: src/Auctionify.Application/Hubs/AuctionHub.cs
Comment (SevenDart) [inline src/Auctionify.Application/Hubs/AuctionHub.cs:1]: In this class you must place only functions, that can be executed from the client side. These commands can't be executed from the client side, only from server, what you've done in the other files like AddBidForLotCommand.

**Review by SevenDart — CHANGES_REQUESTED** [review]

File: src/Auctionify.Application/Features/Lots/Queries/FIlter/FilterLotsResponse.cs
[the remark this answers is not in the document]
Comment (khasanrashidov) [reply src/Auctionify.Application/Features/Lots/Queries/FIlter/FilterLotsResponse.cs:1]: As you see there is no difference between them in terms of what they return, they both inherit from GetAllLots (from BaseQueryModels), but in terms of chosen structure or architecture (CQRS), our queries have their queries and responses,  that is why we need to have their responses even have identical responses in code (but GetAllLots returns all lots, but FilterLots returns filtered lots (based on pointed filter configuration parameters). In general, the code is not duplicate of each other. Because the main response body is in BaseQueryModels.GetAllLots):

image

image

File: src/Auctionify.Application/Features/Users/Commands/RemoveBid/RemoveBidCommandValidator.cs
Comment (khasanrashidov) [reply src/Auctionify.Application/Features/Users/Commands/RemoveBid/RemoveBidCommandValidator.cs:1]: Yeah, cmon)) there is actually one to validate, it says "You can withdraw ONLY YOUR recent bid", it checks the cases when someone tries to withdraw the bid which doesn't belong to him:

Response:
image

Validation Logic:
image

File: src/Auctionify.Application/Hubs/AuctionHub.cs
Comment (khasanrashidov) [reply src/Auctionify.Application/Hubs/AuctionHub.cs:1]: Can you specify the problem, please? I didn't quite get it

File: src/Auctionify.Application/Features/Users/Commands/AddBidForLot/AddBidForLotCommand.cs (hunk 1 above)
Comment (khasanrashidov) [reply src/Auctionify.Application/Features/Users/Commands/AddBidForLot/AddBidForLotCommand.cs]: fixed, grouping is done and hub is authorized:

image

File: src/Auctionify.Application/Features/Lots/Queries/FIlter/FilterLotsResponse.cs
[the remark this answers is not in the document]
Comment (SevenDart) [reply src/Auctionify.Application/Features/Lots/Queries/FIlter/FilterLotsResponse.cs:1]: Ok, got it, thanks for clarification

File: src/Auctionify.Application/Features/Users/Commands/AddBidForLot/AddBidForLotCommand.cs (hunk 1 above)
Comment (SevenDart) [reply src/Auctionify.Application/Features/Users/Commands/AddBidForLot/AddBidForLotCommand.cs]: As I see, in current implementation, when user navigates to another page in the application, they won't be removed from the group and will receive notifications, you should add leaving group method for every time user closes page with lot.

File: src/Auctionify.Application/Hubs/AuctionHub.cs
Comment (SevenDart) [reply src/Auctionify.Application/Hubs/AuctionHub.cs:1]: As we already discussed, you need to place in AuctionHub methods, that can be invoked from the client, not from the server. In your case clients shouldn't invoke methods 'SendBidNotification' or 'SendWithdrawBidNotification' directly, this message will be send during the workflow of your command handler

File: src/Auctionify.Application/Features/Users/Commands/AddBidForLot/AddBidForLotCommand.cs [hunk 4]
```diff
@@ -57,7 +57,9 @@ CancellationToken cancellationToken
 
 			var result = await _bidRepository.AddAsync(bid);
 
-			await _hubContext.Clients.All.SendAsync("ReceiveBidNotification", cancellationToken: cancellationToken);
+			await _hubContext.Clients
+				.Group(request.LotId.ToString())
+				.SendAsync("ReceiveBidNotification", cancellationToken: cancellationToken);
```
Comment (SevenDart) [inline src/Auctionify.Application/Features/Users/Commands/AddBidForLot/AddBidForLotCommand.cs:62]: You should use constants instead of strings for signalR actions, as you did on the client side

File: src/Auctionify.Application/Features/Users/Commands/RemoveBid/RemoveBidCommand.cs [hunk 5]
```diff
@@ -42,10 +42,9 @@ CancellationToken cancellationToken
 
 			await _bidRepository.UpdateAsync(bid!);
 
-			await _hubContext.Clients.All.SendAsync(
-				"ReceiveWithdrawBidNotification",
-				cancellationToken: cancellationToken
-			);
+			await _hubContext.Clients
+				.Group(bid!.LotId.ToString())
+				.SendAsync("ReceiveWithdrawBidNotification", cancellationToken: cancellationToken);
```
Comment (SevenDart) [inline src/Auctionify.Application/Features/Users/Commands/RemoveBid/RemoveBidCommand.cs:47]: same as in AddBidForLotCommand

**Review by SevenDart — CHANGES_REQUESTED** [review]

File: src/Auctionify.Application/Features/Users/Commands/AddBidForLot/AddBidForLotCommand.cs (hunk 1 above)
Comment (khasanrashidov) [reply src/Auctionify.Application/Features/Users/Commands/AddBidForLot/AddBidForLotCommand.cs]: fixed, thank you

**Review by otash8110 — APPROVED** [review]
